nixos/modules/services/web-apps/invidious.nix

```
@@ -7,29 +7,60 @@ let
 
   settingsFile = settingsFormat.generate "invidious-settings" cfg.settings;
 
+  finalHmacKeyFile =
+    if cfg.hmacKeyFile != null
+    then cfg.hmacKeyFile
+    else "/var/lib/invidious/hmac_key";
+  generateHmac = cfg.hmacKeyFile == null;
+
   serviceConfig = {
     systemd.services.invidious = {
       description = "Invidious (An alternative YouTube front-end)";
       wants = [ "network-online.target" ];
       after = [ "network-online.target" ];
       wantedBy = [ "multi-user.target" ];
 
-      script =
-        let
-          jqFilter = "."
-            + lib.optionalString (cfg.database.host != null) "[0].db.password = \"'\"'\"$(cat ${lib.escapeShellArg cfg.database.passwordFile})\"'\"'\""
-            + " | .[0]"
-            + lib.optionalString (cfg.extraSettingsFile != null) " * .[1]";
-          jqFiles = [ settingsFile ] ++ lib.optional (cfg.extraSettingsFile != null) cfg.extraSettingsFile;
-        in
-        ''
-          export INVIDIOUS_CONFIG="$(${pkgs.jq}/bin/jq -s "${jqFilter}" ${lib.escapeShellArgs jqFiles})"
-          exec ${cfg.package}/bin/invidious
-        '';
+      preStart = lib.optionalString generateHmac ''
+        if [[ ! -e "${finalHmacKeyFile}" ]]; then
+          ${pkgs.pwgen}/bin/pwgen 20 1 > "${finalHmacKeyFile}"
+          chmod 0600 "${finalHmacKeyFile}"
+        fi
+      '';
+
+      script = ''
+        configParts=()
+      ''
+      # autogenerated hmac_key
+      + lib.optionalString generateHmac ''
+        configParts+=("$(${pkgs.jq}/bin/jq -R '{"hmac_key":.}' <"${finalHmacKeyFile}")")
+      ''
+      # generated settings file
+      + ''
+        configParts+=("$(cat "${settingsFile}")")
```

This doesn't strictly need `lib.escapeShellArg` because it's a Nix store path, which I'm pretty sure can't contain any sequence that would need to be escaped in bash, but let's do it anyways for consistency with my other suggestions, along with `cat`:

```suggestion
        configParts+=("$(<${lib.escapeShellArg settingsFile})")
```